src/credentials/credentialsUtilities.ts

```diff
@@ -53,3 +58,44 @@ export function notifyUserInvalidCredentials(credentialProviderId: CredentialsPr
 export function hasProfileProperty(profile: Profile, propertyName: string): boolean {
     return !!profile[propertyName]
 }
+
+/**
+ * Attempts to resolve (or refresh) a provider with a 'Cancel' progress message.
+ * User cancellation or timeout expiration will cause rejection.
+ *
+ * @param profile Profile name to display for the progress message
+ * @param provider This can be a Promise that returns Credentials, or void if using 'refresh'
+ * @param timeout How long to wait for resolution without user intervention (default: 5 minutes)
+ *
+ * @returns The resolved Credentials or undefined if the the provider was a 'refresh' Promise
+ */
+export async function resolveProviderWithCancel<T extends AWS.Credentials | void>(
+    profile: string,
+    provider: Promise<T>,
+    timeout: Timeout | number = CREDENTIALS_TIMEOUT
```

Comment:
Would it be easier just to make the caller responsible for passing a Timeout object? Lots of gymnastics here to make that work.

Reply:
I don't think it's that much extra work to handle both a Timeout and a number. Versatility seems like a nice thing to have.